File: bot/src/command/GiveCommand.ts
import Command from "./Command";
import {VkMessage} from "../service/VkMessagesService";

export default class GiveCommand extends Command {
    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command == "give";
    }

    getCommandShortUsage(): string {
        return "/sweet give (credits_amount) (user)";
    }

    chatAdminOnly(peerId: number): boolean {
        return true;
    }

    appCeoOnly(): boolean {
        return true;
    }

    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService } = this.context;

        const firstArg = rawArguments.split(" ")[0];
        const restArgs = rawArguments.substring(firstArg.length + 1);
        const creditsAmount = parseInt(firstArg);
        if (isNaN(creditsAmount)) {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }

        let memberId: number | null;
        if (restArgs.length > 0) {
            memberId = this.extractMemberId(restArgs);
            if (memberId == null) {
                await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
                return;
            }
        } else if (message.forwardedMessages.length == 1) {
            memberId = message.forwardedMessages[0].fromId;
        } else {
            await vkMessagesService.send(message.peerId, this.getUsage());
            return;
        }
        const user = await this.context.vkUsersService.getUser(memberId);
        if (!user) {
            await vkMessagesService.send(message.peerId, `Не могу найти такого пользователя`);
            return;
        }
        await this.context.vkUsersOrmService.addCredits(memberId, creditsAmount);
        await vkMessagesService.send(message.peerId, `✨ Выдаю ${creditsAmount} кредитов`);
    }

    private extractMemberId(text: string): number | null {
        const regex = /\[?(id|club)(\d+)\|?/;
        const match = text.match(regex);
        if (match && !isNaN(+match[2])) {
            return +match[2];
        } else {
            return null;
        }
    }

    private getUsage(): string {
        let result = '';
        result += 'Так пиши:\n';
        result += '/sweet give (credits_amount) (user)\n';
        return result;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/GiveCommand.kt
package space.davids_digital.sweetie.command

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.integration.vk.VkUtils
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.orm.repository.VkUserRepository

@Component
class GiveCommand(
    private val vkMessageService: VkMessageService,
    private val vkUserRepository: VkUserRepository
): Command {
    companion object {
        private val log = LoggerFactory.getLogger(GiveCommand::class.java)
    }

    override fun getNames(): Array<String> {
        return arrayOf("give")
    }

    override fun getUsage(): String {
        return "give (credits_amount) (user)"
    }

    override fun requiresChatAdmin(): Boolean {
        return true
    }

    override fun requiresAppCeo(): Boolean {
        return true
    }

    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        if (rawArguments.isBlank()) {
            return handleHelp(message)
        }
        val arguments = rawArguments.split(" ")
        val creditsAmount = arguments[0].toIntOrNull() ?: return handleHelp(message)
        val rest = arguments.drop(1).joinToString(" ")

        val memberId: Long?
        if (arguments.size >= 2) {
            memberId = VkUtils.extractMemberId(rest)
            if (memberId == null) {
                vkMessageService.send(message.peerId, "Не могу найти участника")
                return
            }
        } else if (message.forwardedMessages.size == 1) {
            memberId = message.forwardedMessages.first().fromId
        } else {
            return handleHelp(message)
        }

        withContext(Dispatchers.IO) {
            vkUserRepository.addCredits(memberId, creditsAmount)
        }
        log.info("User ${message.fromId} gave $creditsAmount credits to $memberId")
        vkMessageService.send(message.peerId, "✨ Выдаю $creditsAmount кредитов")
    }

    private fun handleHelp(message: VkMessageModel) {
        vkMessageService.send(message.peerId, """
            Так пиши:
            /sweet give (credits_amount) (user)
            """.trimMargin()
        )
    }
}